crates/nu-command/src/network/port.rs

```diff
@@ -121,25 +120,200 @@ fn get_free_port(
+    }
+
+    const _: () = assert!(align_of::<MIB_TCPTABLE2>() == 4);
+
+    impl TcpTable {
+        fn new() -> Result<Box<Self>, WIN32_ERROR> {
+            let mut size = 0;
+            let size_pointer: *mut u32 = &mut size;
+
+            // SAFETY:
+            // - Passing a null table pointer queries the required size (documented behavior).
+            // - `size_pointer` is a valid, non-null out pointer.
+            // - We expect `ERROR_INSUFFICIENT_BUFFER` so that `size` is written.
+            let ret_code = unsafe { GetTcpTable2(None, size_pointer, false) };
+            assert_eq!(WIN32_ERROR(ret_code), ERROR_INSUFFICIENT_BUFFER);
+
+            // SAFETY:
+            // - Alignment is 4: non-zero and a power of two.
+            // - `size` comes from the API and is expected to be reasonable for allocation.
+            let layout = unsafe {
+                Layout::from_size_align_unchecked(size as usize, align_of::<MIB_TCPTABLE2>())
+            };
+
+            // IMPORTANT: This allocation must be freed or transferred to ownership before leaving this scope.
+            // SAFETY: `layout` has non-zero size (at least 4 for one u32).
+            let ptr = unsafe { alloc(layout) as *mut MIB_TCPTABLE2 };
+            assert!(!ptr.is_null());
+
+            // SAFETY:
+            // - `ptr` is non-null, properly aligned, and points to `size` bytes.
+            // - `size_pointer` still points to `size` from the first call.
+            let ret_code = unsafe { GetTcpTable2(Some(ptr), size_pointer, false) };
+            let ret_code = WIN32_ERROR(ret_code);
+            if ret_code != NO_ERROR {
+                // SAFETY:
+                // - `ptr` was allocated with `alloc(layout)` in this function.
+                // - Using the same `layout` to deallocate is correct.
+                unsafe { dealloc(ptr as *mut u8, layout) };
+                return Err(ret_code);
+            }
+
+            // SAFETY: `GetTcpTable2` returned `NO_ERROR`, so the header at `ptr` is initialized.
+            let header = unsafe { &*ptr };
+
+            // SAFETY:
+            // - Memory at `ptr` came from the global allocator and is initialized.
+            // - `TcpTable` is #[repr(C)] and layout-compatible with `MIB_TCPTABLE2` plus trailing rows.
+            // - We build a slice fat pointer only to carry the length; we do not dereference the slice itself here.
+            // - Casts between slice DSTs preserve the length metadata:
+            //     https://github.com/rust-lang/unsafe-code-guidelines/issues/288
+            //     https://github.com/rust-lang/reference/pull/1417
+            // - Casting to `*mut TcpTable` preserves that metadata for our DST.
+            // - `Box::from_raw` takes ownership and will free via the same allocator.
+            let table = unsafe {
+                let ptr = ptr::slice_from_raw_parts_mut(ptr, header.dwNumEntries as usize);
+                Box::from_raw(ptr as *mut TcpTable)
```

Comment:
That is certainly a spicy bit with custom DSTs. Looks wrong but `ptr` defined as a slice is never read as such only created to cast to create the DST without the nightly `Pointee::Metadata`

https://doc.rust-lang.org/std/ptr/trait.Pointee.html#associatedtype.Metadata